tensor.go

```diff
@@ -6,45 +6,87 @@ package gotorch
 // #include "cgotorch.h"
 import "C"
 
+import (
+	"runtime"
+	"sync"
+)
+
+var (
+	cTensorFinalizersWG = &sync.WaitGroup{}
+	gcPrepared          = false
+)
+
+func setTensorFinalizer(t *C.Tensor) {
+	if gcPrepared {
+		cTensorFinalizersWG.Add(1)
+	}
+	runtime.SetFinalizer(t, func(t *C.Tensor) {
+		go func() {
+			C.Tensor_Close(*t)
+			cTensorFinalizersWG.Done()
+		}()
+	})
+}
+
+// PrepareGC should be called right before a train/predict loop
+func PrepareGC() {
+	gcPrepared = true
```

Comment:
Why do we need gcPrepared?

Reply:
We have to use this to mark variables with different lifecycles, i. e. variables live in and out of the train loop.